homeassistant/components/goodwe/sensor.py

```diff
@@ -0,0 +1,266 @@
+"""Support for GoodWe inverter via UDP."""
+from goodwe import SensorKind
+import voluptuous as vol
+
+from homeassistant.components.sensor import (
+    STATE_CLASS_MEASUREMENT,
+    STATE_CLASS_TOTAL_INCREASING,
+    SensorEntity,
+    SensorEntityDescription,
+)
+from homeassistant.const import (
+    DEVICE_CLASS_BATTERY,
+    DEVICE_CLASS_CURRENT,
+    DEVICE_CLASS_ENERGY,
+    DEVICE_CLASS_POWER,
+    DEVICE_CLASS_TEMPERATURE,
+    DEVICE_CLASS_VOLTAGE,
+    ELECTRIC_CURRENT_AMPERE,
+    ELECTRIC_POTENTIAL_VOLT,
+    ENERGY_KILO_WATT_HOUR,
+    FREQUENCY_HERTZ,
+    POWER_WATT,
+    TEMP_CELSIUS,
+)
+from homeassistant.helpers import config_validation as cv, entity_platform
+from homeassistant.helpers.entity import DeviceInfo
+from homeassistant.helpers.update_coordinator import CoordinatorEntity
+
+from .const import DOMAIN, KEY_COORDINATOR, KEY_INVERTER
+
+# Service related constants
+SERVICE_SET_WORK_MODE = "set_work_mode"
+ATTR_WORK_MODE = "work_mode"
+SET_WORK_MODE_SERVICE_SCHEMA = vol.Schema(
+    {
+        vol.Required(ATTR_WORK_MODE): cv.positive_int,
+    }
+)
+SERVICE_SET_ONGRID_BATTERY_DOD = "set_ongrid_battery_dod"
+ATTR_ONGRID_BATTERY_DOD = "ongrid_battery_dod"
+SET_ONGRID_BATTERY_DOD_SERVICE_SCHEMA = vol.Schema(
+    {
+        vol.Required(ATTR_ONGRID_BATTERY_DOD): cv.positive_int,
+    }
+)
+SERVICE_SET_GRID_EXPORT_LIMIT = "set_grid_export_limit"
+ATTR_GRID_EXPORT_LIMIT = "grid_export_limit"
+SET_GRID_EXPORT_LIMIT_SERVICE_SCHEMA = vol.Schema(
+    {
+        vol.Required(ATTR_GRID_EXPORT_LIMIT): cv.positive_int,
+    }
+)
+
+_MAIN_SENSORS = (
+    "house_consumption",
+    "active_power",
+    "battery_soc",
+    "e_day",
+)
+
+_ICONS = {
+    SensorKind.PV: "mdi:solar-power",
+    SensorKind.AC: "mdi:power-plug-outline",
+    SensorKind.UPS: "mdi:power-plug-off-outline",
+    SensorKind.BAT: "mdi:battery-high",
+    SensorKind.GRID: "mdi:transmission-tower",
+}
+
+_DEVICE_CLASSES = {
+    "A": DEVICE_CLASS_CURRENT,
+    "V": DEVICE_CLASS_VOLTAGE,
+    "W": DEVICE_CLASS_POWER,
+    "kWh": DEVICE_CLASS_ENERGY,
+    "C": DEVICE_CLASS_TEMPERATURE,
+    "Hz": DEVICE_CLASS_VOLTAGE,
+}
+
+_STATE_CLASSES = {
+    "A": STATE_CLASS_MEASUREMENT,
+    "V": STATE_CLASS_MEASUREMENT,
+    "W": STATE_CLASS_MEASUREMENT,
+    "kWh": STATE_CLASS_TOTAL_INCREASING,
+    "C": STATE_CLASS_MEASUREMENT,
+    "Hz": STATE_CLASS_MEASUREMENT,
+}
+
+_UNITS = {
+    "A": ELECTRIC_CURRENT_AMPERE,
+    "V": ELECTRIC_POTENTIAL_VOLT,
+    "W": POWER_WATT,
+    "kWh": ENERGY_KILO_WATT_HOUR,
+    "C": TEMP_CELSIUS,
+    "Hz": FREQUENCY_HERTZ,
+}
+
+
+def _get_sensor_description(sensor):
+    """Create entity description for specified inverter sensor."""
+    desc = SensorEntityDescription(
```

Comment:
I think `SensorEntityDescription` should be constants, and not generated on the fly.
Then `_get_sensor_description` "picks" the correct SensorEntityDescription from the list.

eg.
```python
DESCRIPTIONS = {
    "pv": SensorEntityDescription(...),
    "ac": SensorEntityDescription(...),
    ...
}
```

Reply:
Well, the problem is there is not a single fixed set of sensors and there are lots of them.
Set of sensors heavily depends on specific PV inverter family, model and enabled capabilities.
There are inverters with just like ~20 sensors and there are inverters exposing 145 sensors.

I see no added value having superset of them hard-coded in HA sources, when all relevant sensor entity characteristics are dynamically provided by the underlying library via its unit and sensorKind attributes.
That's whey the _get_sensor_description, to avoid those useless ~160 constants.

I understand how and why it is done in other integrations, it makes sense there, since probably 95% of them have fixed, pre-defined set of sensors. But as mentioned above, it's not the case in these PV inverters.

Reply:
You wouldn't need to create 160 constants.

You would create a single constant for the (say) 50 `power` sensors. The shared items (`device_class`, `icon`, `state_class`, `unit_of_measurement`) are inside the EntityDescription constant.
And then all the specific items (name, unique_id, etc.) use the `_attr_` form instead
```python
self._attr_name = ...
self._attr_unique_id = ...
```

Reply:
I see, will try.

Reply:
I'm watching this to learn. How I'm understanding the problem, there would need to be ~160 EntityDescriptions. Isn't the `key` field the main thing to help pull the data from the integration lib? Don't slow down because of my comment... I'm catching up on my understanding and will continue to watch/think.

Reply:
`key` doesn't have to be unique, and doesn't have to be used. In my opinion it should have been made optional to begin with. Originally it wasn't even there and those who required it created a dict {key1: entitydescription1, key2: entitydescription2}.
If I am mistaken and the key is automatically used somewhere in the base classes then please point it out to me.

Reply:
Done the suggested way.